compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp

```diff
@@ -266,15 +278,238 @@ static LogicalResult replaceAllStoresWithTiledVersion(
+    if (!ubOp) {
+      // TODO: handle dynamic cases by adding more code to check the
+      // conditions.
+      return success();
+    }
+    auto ub = ubOp.value();
+    auto tileSize = std::get<1>(it);
+    if (ub % tileSize != 0) {
+      isAlreadyMultiple = false;
+    }
+    if (ub < tileSize) {
+      return success();
+    }
+  }
+
+  if (isAlreadyMultiple) return success();
+
+  auto minSizeOps = collectTileSizeMinOps(tilingResult.boundedSizesForLoops);
+
+  LLVM_DEBUG({ tilingResult.dump(); });
+
+  PatternRewriter::InsertionGuard guard(rewriter);
+  auto distLoop0 = distLoops[0];  // the outermost loop
+  auto loc = distLoop0.getLoc();
+  rewriter.setInsertionPoint(distLoop0);
+
+  // create a condition for scf.if
+  Value cond;
+  SmallVector<Value> constantOps;  // ConstantIndexOps for tile sizes
+  for (unsigned i = 0, e = distLoops.size(); i != e; ++i) {
+    auto tileSize = (*tileSizes)[i];
+    if (tileSize == 0 || tileSize == 1) {
+      constantOps.push_back(Value());
+      continue;
+    }
+
+    // clone the minSize op in the loop and place it before scf.if
+    auto minOp = minSizeOps[i];
+    auto distLoop = distLoops[i];
+    // clone the lower bound and put before the nested loops.
+    BlockAndValueMapping mapperForLB;
+    Operation *lb =
+        rewriter.clone(*distLoop.getLowerBound().getDefiningOp(), mapperForLB);
+
+    // Clone the affine min op for the dynamic size in the current loop and
+    // place it before the nested loops. The induction variable is replaced by
+    // the cloned lower-bound above.
+    BlockAndValueMapping mapperForIV;
+    auto iv = distLoop.getInductionVar();
+    mapperForIV.map(iv, lb->getResult(0));
+    Value size =
+        rewriter.clone(*minOp.getOperation(), mapperForIV)->getResult(0);
+
+    // Generate a compare op that checks the dynamic size is equal to the
+    // constant main tile size.
+    Value constant = rewriter.create<arith::ConstantIndexOp>(loc, tileSize);
+    constantOps.push_back(constant);
+    Value cmp = rewriter.create<arith::CmpIOp>(loc, arith::CmpIPredicate::eq,
+                                               size, constant);
+    cond = cond ? rewriter.create<arith::AndIOp>(loc, cond, cmp) : cmp;
+  }
+
+  // generate scf.if %cond then { scf.for ... } else { scf.for ... }
+  auto ifOp = rewriter.create<scf::IfOp>(loc, cond, /*withElseRegion=*/true);
+  ifOp.getElseBodyBuilder().clone(*distLoop0.getOperation());
+
+  // Specialize the size operations (affine min) in each for loop.
+  for (int i = 0, e = distLoops.size(); i != e; ++i) {
+    auto minOp = minSizeOps[i];
+    if (!minOp) {
+      assert((*tileSizes)[i] == 0 || (*tileSizes)[i] == 1);
+      continue;
+    }
+    LLVM_DEBUG({
+      llvm::dbgs() << "Replacing ";
+      minOp.dump();
+      llvm::dbgs() << "with ";
+      constantOps[i].dump();
+    });
+    rewriter.replaceOp(minOp, constantOps[i]);
+  }
+  ifOp.getThenBodyBuilder().clone(*distLoop0.getOperation());
```

Comment:
You can probably do this without a clone.

Reply:
You mean moving? I'd love to know how.

Reply:
You can use https://github.com/llvm/llvm-project/blob/8d2a447bf92700e576e037c5f6495f74153cf5e9/mlir/include/mlir/IR/PatternMatch.h#L404 and its friends.

Reply:
Cool! Let me try!

Reply:
Actually there is `moveBefore()`. `inlineRegionBefore()` only moves `Region`.

Reply:
Done!